### PR TITLE
Feature/parameterized dist

### PR DESCRIPTION
**Pull request by ccnobbli**
#186 implemented tests in new format, and some in the old format.

### CODE REVIEW

**Comment by jpcampb2** [comment]
@ccnobbli Really excited for this PR, I think it'll be a nice expectation to be able to clearly express. Thank you.

File: great_expectations/dataset/base.py [hunk 1]
```diff
@@ -1980,6 +1980,81 @@ def expect_column_values_to_match_json_schema(self,
...
+                The threshold p-value for a passing test. Default is 0.05
+            params (dict) : \
+                A dictionary of shape parameters that describe the disrtibution you want to test the data against.\
+                If not provided, the parameters will be attempted to be constructed using the column data and \
+                the specified distribution. Include key values specific to the distribution from the appropriate scipy \
+                distribution CDF function. \
+                See https://docs.scipy.org/doc/scipy/reference/stats.html#continuous-distributions
+
+        Other Parameters:
+            output_format (str or None): \
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: Small tweak needed here: we renamed output_format to result_format in the new docs. Grab the "other parameters" section from an expectation now to update, please.

File: great_expectations/dataset/base.py [hunk 2]
```diff
@@ -1980,6 +1980,81 @@ def expect_column_values_to_match_json_schema(self,
...
+            These fields in the result object are customized for this expectation:
+            ::
+
+                {
+                    "details":
+                        "expected_params" (dict): The specified or inferred parameters of the distribution to test against
+                        "ks_results" (dict): The raw result of stats.kstest()
+                }
+
+            * The Kolmogorov-Smirnov test's null hypothesis is that the column is similar to the provided distribution.
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py:2045]: 👍 Great, clear docs.

File: great_expectations/dataset/util.py [hunk 3]
```diff
@@ -297,3 +297,253 @@ def continuous_partition_data(data, bins='auto', n_bins=10):
...
+            "beta": (float),
+            "min": (float),
+            "max": (float),
+            "df": (float)
+        }
+
+        See: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kstest.html#scipy.stats.kstest
+    """
+
+    if params is None:
```
Comment (jpcampb2) [inline great_expectations/dataset/util.py:329]: I think this is (perhaps unavoidably because we inherit the confusion from scipy) confusing, but I say let's go with it for now.

File: great_expectations/dataset/util.py [hunk 4]
```diff
@@ -297,3 +297,253 @@ def continuous_partition_data(data, bins='auto', n_bins=10):
...
+        raise AttributeError("Unsupported distribution type. Please refer to Great Expectations Documentation")
+
+    params['loc'] = params.get('loc', 0)
+    params['scale'] = params.get('scale', 1)
+
+    return params
+
+def _scipy_distribution_positional_args_from_dict(distribution, params):
+    """Helper function that returns positional arguments for a scipy distribution using a dictionary of parameters.
+
```
Comment (jpcampb2) [inline great_expectations/dataset/util.py]: Can you link to the scipy docs to show where these magic orders come from?

File: tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+      "norm_std" : [-0.761837213353,-0.331616898194,-1.75131542992,0.628894110773,0.282501864129,-1.33813943,-0.50060684963,0.121645029892,1.70832347381,-0.970999448321,-0.619332343444,-0.726708131791,1.22165541672,0.503699288341,-1.3878740774,0.204851419543,0.603705215945,0.545680308693,0.235477019441,0.111834993822,-1.2515037504,-2.94934349814,0.634634160585,0.124157016111,1.29762248959,-1.68693341116,1.08953904655,2.06088173968,-0.241235326269,-0.94787218032,0.676294028923,-0.653356162094,-0.652295297944,0.528827604205,0.357793249335,0.188649359732,0.869416879035,-0.0506674481438,-0.71636457461,-0.103258720839,-1.1410365794,-0.5
```
Comment (jpcampb2) [inline tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json]: All of these totally make sense, but it seems odd to test such wildly different distributions. Even for people scanning our test suite, I think it would be clearer to just use params like N(1,1) here (which will still fail KS with a sample of 500 drawn from N(0,1))

Could you please tighten your negative cases throughout to just show that more realistic parallelism?

File: tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
...2334,0.0145528928139,0.028287672854,0.00161394683637,0.035612390441,0.0310593922782,0.0208640386703,0.0606897615431,0.00155320238875,0.00703216101611,0.0258813696042,0.092212785197,0.215359740123,0.00211753756519,0.119870782455,0.109345596481,0.0199830265157,0.0369247294863,0.12922036566,0.125047275129,0.0300346930819,0.0164191975644,0.0484312006994,0.00036731162446,0.275721532273,0.00849742412309,0.102615988164,0.077110641212,0.0425222319674,0.00883196654238,0.00107229664453,0.217989672828,0.000139638732251,0.0256345246264,0.0565471555872,0.0166802315519,0.0571633494295,0.00171391540273,0.000433437968232,0.00933279310548,0.00667171052542,0.0371666643478,0.00537213525143,0.0461674115441,0.01010732596,0.00435954509591,0.00205961774847,0.000200079567974,0.000838080295037,0.00844047131635,0.000971570628295,0.0063739668837,0.00805974493667,0.0168097093896,0.0428648759884,0.00561973605881,0.0098940830938,0.13830882269,0.11308181244,0.109262608978,0.0131659663175,0.16010707566,0.0147394038013,3.81000767445e-05,0.0588317876967,0.0487327085473,0.214498999773,0.0210924377134,0.0663552773565,0.0969217570186,0.000204338801391,0.00147181989751,0.0114056270752,0.00143897011757,0.0135320064798],
```
Comment (jpcampb2) [inline tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json]: It looks like we're only using the _std version of the datasets. Do we need them all?

File: tests/test_pandas_dataset.py [hunk 7]
```diff
@@ -4,6 +4,7 @@
 import json
 import numpy as np
 import datetime
+import scipy.stats as stats
```
Comment (jpcampb2) [inline tests/test_pandas_dataset.py]: This can be pulled back again, right?

File: tests/test_sets/fixed_distributional_test_dataset.csv [hunk 8]
```diff
@@ -0,0 +1,501 @@
+beta,beta_std,chi2,chi2_std,exponential,exponential_std,gamma,gamma_std,norm,norm_std,poisson,poisson_std,uniform,uniform_std
```
Comment (jpcampb2) [inline tests/test_sets/fixed_distributional_test_dataset.csv]: Now that the tests are in the new json format, is this file needed?

File: tests/test_sets/fixed_distributional_test_dataset.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
...629,5.29209873314,5.37324702275,5.00214663987,5.72588480981,5.01607694333,7.89194085497,5.56651398908,5.0228299548,5.48817828506,7.876897102,5.13818010279,5.61058224818,5.3172239619,6.39136992112,5.01370811787,5.02506106002,5.49581934891,5.13769593786,6.49548844493,5.36836022876,5.20207116155,5.20978364036,5.81889314196,5.01181447295,7.20812760853,5.02394728416,5.81278988953,5.00153016661,5.00177767271,5.11118798548,5.0007402966,5.02280054287,5.03341758946,6.28060848763,5.02377637811,5.89602305564,5.61092446091,5.27649878517,5.68406005403,6.05602295961,5.38068063436,5.05924695009,5.23858897507,5.05209761333,5.06463295681,5.06301948811,5.06327128273,5.22909870866,5.51075923844,5.21891194219,5.08883096523,5.09895239096,5.66941984233,5.01681794776,5.00597357028,6.87556724012,5.13018807678,5.36623298782,5.66410515544,5.22302758816,5.12770982759,5.08566544477,5.55532066532,5.70621282514,5.05093746427,5.34108431674,5.00470684237,6.23171040566,5.70887956287,5.28900035873,5.02438488984,5.61523467551,5.06419435981,6.32495050302,6.53163282487,5.2174067149,5.03976523607,5.41343155571,5.29689448172,6.11928120444,5.26486548936,5.60696900246,5.058876487,5.8957845307,5.98635366819,5.25057476532],
```
Comment (jpcampb2) [inline tests/test_sets/fixed_distributional_test_dataset.json:1]: same as above...this is now in the json file, right?

File: tests/test_utils.py [hunk 10]
```diff
@@ -140,6 +141,7 @@ def evaluate_json_test(dataset, expectation_type, test):
     """
 
     dataset.set_default_expectation_argument('result_format', 'COMPLETE')
+    dataset.set_default_expectation_argument('catch_exceptions', True)
```
Comment (jpcampb2) [inline tests/test_utils.py]: To see the impact of this change, I added the following to evaluate_json_test:
```
        if not ('success' in test['out'] or 'traceback_substring' in test['out']):
            warnings.warn("This test does not test success or traceback_substring. Is it defined improperly?")

        if result['exception_info']['raised_exception'] is True and 'traceback_subsring' not in test['out']:
            warnings.warn("This test did not test a traceback substring, but the expectation raised an exception.")
```

That did trigger extra warnings, so I think we need to double check the tests throughout. Probably not an issue with your PR, but I want to look into it while you make the other changes.

**Review by jpcampb2 — CHANGES_REQUESTED** [review]
Great stuff! Really excited. Inline comments regarding changes.

File: tests/test_pandas_dataset.py (hunk 7 above)
Comment (ccnobbli) [reply tests/test_pandas_dataset.py]: Yup, I went ahead and removed it.

File: tests/test_sets/fixed_distributional_test_dataset.csv (hunk 8 above)
Comment (ccnobbli) [reply tests/test_sets/fixed_distributional_test_dataset.csv]: It is being used currently for the test_dataset_util tests in test_infer_distribution_parameters and test_validate_distribution_parameters... but if you know of a way to load in dataset from a JSON file instead of csv, than these are not needed!

**Comment by ccnobbli** [comment]
Do I need to re-pull request the newest push from this branch after making a few edits? or is it automatic? @jpcampb2 @abegong

File: tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json (hunk 6 above)
Comment (ccnobbli) [reply tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json]: Most of the _std versions aren't used. I envisioned using them to make sure that the expectation automatically uses standard loc and scale parameters if not provided. It only needs to be tested on one distribution though, because of the simple code path.

**Comment by jpcampb2** [comment]
It's automatic. Just flag in comments if you're ready for the updates to be reviewed again, and ideally respond to my inline comments in the code. The ones you already did look great.

File: great_expectations/dataset/base.py (hunk 1 above)
Comment (ccnobbli) [reply great_expectations/dataset/base.py]: Went ahead and copied over the boilerplate documentation from another expectation. It should match the current nomenclature.

**Comment by ccnobbli** [comment]
Great, I think it's ready to reviewed @jpcampb2, and sorry for my ignorance but when you say 'inline comments in the code' do you mean literal comments that you put in the code or just the github conversations below the code blocks?

File: great_expectations/dataset/util.py (hunk 3 above)
Comment (ccnobbli) [reply great_expectations/dataset/util.py:329]: Am totally up for any suggestions. I'll commit some changes that may clear it up a bit more.

**Comment by jpcampb2** [comment]
Just the github comments! :)

File: tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json (hunk 5 above)
Comment (ccnobbli) [reply tests/column_aggregate_expectations/expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than.json]: Tightened like a screw - it works even better than I thought it would.

**Comment by ccnobbli** [comment]
Ok, there were comments I did not see before, I addressed all of them and should be ready for final review!

**Review by jpcampb2 — APPROVED** [review]
